### PR TITLE
chore: add a class to handle aggreggation queries

### PR DESCRIPTION
**Pull request by gastonfournier**
## About the changes
We have many aggregation queries that run on a schedule: https://github.com/Unleash/unleash/blob/f63496d47f6808a246732f5ddee9b9136e23eca7/src/lib/metrics.ts#L714-L719

These staticCounters are usually doing db query aggregations that traverse tables and we run all of them in parallel: https://github.com/Unleash/unleash/blob/f63496d47f6808a246732f5ddee9b9136e23eca7/src/lib/metrics.ts#L410-L412

This can add strain to the db. This PR suggests a way of handling these queries in a more structured way, allowing us to run them sequentially (therefore spreading the load): https://github.com/Unleash/unleash/blob/f02fe87835c30d77ff8b6755268e123585820511/src/lib/metrics-gauge.ts#L38-L40

As an additional benefit, we get both the gauge definition and the queries in a single place: https://github.com/Unleash/unleash/blob/f02fe87835c30d77ff8b6755268e123585820511/src/lib/metrics.ts#L131-L141

This PR only tackles 1 metric, and it only focuses on gauges to gather initial feedback. The plan is to migrate these metrics and eventually incorporate more types (e.g. counters)

### CODE REVIEW

File: src/lib/metrics.ts [hunk 1]
```diff
@@ -715,7 +720,6 @@ export default class MetricsMonitor {
             collectStaticCounters.bind(this),
             hoursToMilliseconds(2),
             'collectStaticCounters',
-            0, // no jitter
```
Comment (gastonfournier) [inline src/lib/metrics.ts:718]: Allowing these statistics to run with jitter as they are not needed at startup

File: src/lib/metrics.ts [hunk 2]
```diff
@@ -125,11 +127,19 @@ export default class MetricsMonitor {
...
+        dbMetrics.registerGaugeDbMetric({
             name: 'max_feature_strategies',
             help: 'Maximum number of strategies in one feature',
             labelNames: ['feature'],
+            query: () =>
+                stores.featureStrategiesReadModel.getMaxFeatureStrategies(),
+            map: (result) => ({
+                count: result.count,
+                labels: { feature: result.feature },
+            }),
```
Comment (gastonfournier) [inline src/lib/metrics.ts:140]: Gauge and query are now declared together

File: src/lib/metrics.ts [hunk 3]
```diff
@@ -408,7 +419,6 @@ export default class MetricsMonitor {
                     instanceOnboardingMetrics,
                     projectsOnboardingMetrics,
                 ] = await Promise.all([
```
Comment (sjaanus) [inline src/lib/metrics.ts:421]: We do not need to run these in parallel. I would be to get rid of promise all.

File: src/lib/metrics-gauge.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    labelNames: string[];
+    query: Query<T>;
+    map: MapResult<T>;
+};
+
+export class DbMetricsMonitor {
+    private tasks: (() => Promise<void>)[] = [];
+    private gauges: Record<string, Gauge<string>> = {};
+
+    constructor() {}
```
Comment (nunogois) [inline src/lib/metrics-gauge.ts:22]: Are we planning on doing something with the constructor? It not, can we omit it?

File: src/lib/metrics-gauge.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import { createGauge, type Gauge } from './util/metrics';
+
+type RestrictedRecord<T extends string[]> = Record<T[number], string>;
```
Comment (nunogois) [inline src/lib/metrics-gauge.ts:3]:
```suggestion
type RestrictedRecord<T extends readonly string[]> = Record<T[number], string>;
```

File: src/lib/metrics.ts (hunk 3 above)
Comment (gastonfournier) [reply src/lib/metrics.ts:421]: Exactly, we are going to do that gradually. When we register a new metric into this new implementation, the new implementation will run them sequentially

File: src/lib/metrics-gauge.ts (hunk 4 above)
Comment (gastonfournier) [reply src/lib/metrics-gauge.ts:22]: I went back and forth with adding some dependencies to it (like timer) and it just remained there... I think this is just details we can fix later

File: src/lib/metrics-gauge.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    name: string;
+    help: string;
+    labelNames: string[];
+    query: Query<T>;
+    map: MapResult<T>;
+};
+
+export class DbMetricsMonitor {
+    private tasks: (() => Promise<void>)[] = [];
+    private gauges: Record<string, Gauge<string>> = {};
```
Comment (nunogois) [inline src/lib/metrics-gauge.ts:20]: Up to you but I would maybe consider making these a `Set` and a `Map` respectively.

**Review by nunogois — APPROVED** [review]
Super nice!

**Pull request merged**